File: backend/app/routes/desktop_commander.py
from __future__ import annotations

import asyncio
from typing import Optional

from fastapi import APIRouter
from pydantic import BaseModel

from app.services import desktop_commander as dc

router = APIRouter(prefix="/desktop-commander", tags=["desktop-commander"])


class FileReadRequest(BaseModel):
    path: str
    offset: int = 0
    length: Optional[int] = None
    max_bytes: Optional[int] = None


class FileWriteRequest(BaseModel):
    path: str
    content: str
    mode: str = "rewrite"
    encoding: str = "utf-8"


class EditBlockRequest(BaseModel):
    file_path: str
    old_string: str
    new_string: str


class DirectoryRequest(BaseModel):
    path: str
    depth: int = 1


class CreateDirRequest(BaseModel):
    path: str


class MoveFileRequest(BaseModel):
    source: str
    destination: str


class FileInfoRequest(BaseModel):
    path: str


class SearchFilesRequest(BaseModel):
    query: str
    path: Optional[str] = None
    max_results: int = 20
    pattern: Optional[str] = None
    recursive: bool = True


class ExecuteCommandRequest(BaseModel):
    command: str
    shell: str = "powershell"
    timeout_ms: int = 30000
    cwd: Optional[str] = None


class DesktopCommanderConfigPatchRequest(BaseModel):
    allowed_directories: Optional[list[str]] = None
    blocked_commands: Optional[list[str]] = None
    max_read_lines: Optional[int] = None
    max_write_lines: Optional[int] = None


@router.get("/health")
async def desktop_commander_health():
    return dc.get_status()


@router.get("/config")
async def desktop_commander_config():
    return dc.get_config()


@router.patch("/config")
async def patch_desktop_commander_config(req: DesktopCommanderConfigPatchRequest):
    return dc.update_config(
        allowed_directories=req.allowed_directories,
        blocked_commands=req.blocked_commands,
        max_read_lines=req.max_read_lines,
        max_write_lines=req.max_write_lines,
    )


@router.post("/read-file")
async def read_file(req: FileReadRequest):
    result = dc.dc_read_file(req.path, req.max_bytes)
    if result.get("success") and req.offset > 0:
        lines = result.get("content", "").splitlines()
        end = req.offset + req.length if req.length else len(lines)
        selected = lines[req.offset:end]
        result["content"] = "\n".join(selected)
        result["lines_read"] = len(selected)
        result["total_lines"] = len(lines)
        result["offset"] = req.offset
    elif result.get("success"):
        lines = result.get("content", "").splitlines()
        limit = req.length or len(lines)
        selected = lines[:limit]
        result["content"] = "\n".join(selected)
        result["lines_read"] = min(limit, len(lines))
        result["total_lines"] = len(lines)
        result["offset"] = 0
    return result


@router.post("/write-file")
async def write_file(req: FileWriteRequest):
    return dc.dc_write_file(req.path, req.content, req.encoding, req.mode)


@router.post("/edit-block")
async def edit_block(req: EditBlockRequest):
    return dc.dc_edit_block(req.file_path, req.old_string, req.new_string)


@router.post("/list-directory")
async def list_directory(req: DirectoryRequest):
    return dc.dc_list_directory(req.path, req.depth)


@router.post("/create-directory")
async def create_directory(req: CreateDirRequest):
    return dc.dc_create_directory(req.path)


@router.post("/move-file")
async def move_file(req: MoveFileRequest):
    return dc.dc_move_file(req.source, req.destination)


@router.post("/get-file-info")
async def get_file_info(req: FileInfoRequest):
    return dc.dc_get_file_info(req.path)


@router.post("/search-files")
async def search_files(req: SearchFilesRequest):
    query = req.query or (req.pattern or "").replace("*", "").replace(".", " ").strip()
    return dc.dc_search_files(query, req.path, req.max_results)


@router.post("/execute-command")
async def execute_command(req: ExecuteCommandRequest):
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(
        None,
        lambda: dc.dc_execute_command(
            req.command,
            shell=req.shell,
            timeout_ms=req.timeout_ms,
            cwd=req.cwd,
        ),
    )


@router.get("/system-info")
async def system_info():
    return dc.dc_get_system_info()

File: src/lib/desktop-commander.ts
import { API_BASE_URL } from './api';

const DC = `${API_BASE_URL}/desktop-commander`;

export interface DCConfig {
  allowed_directories: string[];
  blocked_commands: string[];
  max_read_lines: number;
  max_write_lines: number;
  home: string;
  version: string;
  default_shell?: string;
  path_separator?: string;
  platform?: string;
  enabled?: boolean;
  ready?: boolean;
  description?: string;
}

export interface DCFileResult {
  success: boolean;
  path?: string;
  content?: string;
  size_bytes?: number;
  truncated?: boolean;
  encoding?: string;
  total_lines?: number;
  lines_read?: number;
  offset?: number;
  description?: string;
}

export interface DCWriteResult {
  success: boolean;
  path?: string;
  bytes_written?: number;
  description?: string;
}

export interface DCDirectoryEntry {
  name: string;
  type: 'file' | 'directory' | 'unknown';
  size_bytes?: number | null;
  modified?: number;
  extension?: string | null;
  error?: string;
}

export type DCDirEntry = DCDirectoryEntry;

export interface DCListResult {
  success: boolean;
  path?: string;
  items?: DCDirectoryEntry[];
  total?: number;
  description?: string;
}

export interface DCSearchEntry {
  path: string;
  name: string;
  snippet?: string;
}

export interface DCSearchResult {
  success: boolean;
  results?: DCSearchEntry[];
  description?: string;
}

export interface DCCommandResult {
  success: boolean;
  exit_code?: number;
  stdout?: string;
  stderr?: string;
  timed_out?: boolean;
  command?: string;
  description?: string;
}

export interface DCFileInfo {
  success: boolean;
  path?: string;
  type?: 'file' | 'directory';
  size?: number;
  created?: number;
  modified?: number;
  permissions?: string;
  line_count?: number;
  description?: string;
}

export interface DCSystemInfoResult {
  success: boolean;
  os: string;
  os_version: string;
  hostname: string;
  cpu_count: number;
  cpu_percent: number;
  memory_total_gb: number;
  memory_used_gb: number;
  memory_percent: number;
  disk_total_gb: number;
  disk_free_gb: number;
  disk_percent: number;
  home_dir: string;
  description: string;
}

async function post<T>(endpoint: string, body: unknown): Promise<T> {
  const response = await fetch(`${DC}/${endpoint}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: `HTTP ${response.status}` }));
    throw new Error(err.detail || `${endpoint} failed: ${response.status}`);
  }

  return response.json();
}

async function get<T>(endpoint: string): Promise<T> {
  const response = await fetch(`${DC}/${endpoint}`);
  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: `HTTP ${response.status}` }));
    throw new Error(err.detail || `${endpoint} failed: ${response.status}`);
  }
  return response.json();
}

export async function getDCConfig(): Promise<DCConfig> {
  try {
    return await get<DCConfig>('config');
  } catch {
    const health = await get<{ ready?: boolean; description?: string }>('health');
    return {
      allowed_directories: [],
      blocked_commands: [],
      max_read_lines: 2000,
      max_write_lines: 2000,
      home: '',
      version: '1.0.0',
      ready: Boolean(health.ready),
      description: health.description,
    };
  }
}

export async function checkDCHealth(): Promise<boolean> {
  try {
    const config = await getDCConfig();
    return config.enabled !== false && config.ready !== false;
  } catch {
    return false;
  }
}

export async function readFile(path: string, offset = 0, length?: number): Promise<DCFileResult> {
  return post<DCFileResult>('read-file', { path, offset, length });
}

export async function writeFile(
  path: string,
  content: string,
  mode: 'rewrite' | 'append' = 'rewrite',
): Promise<DCWriteResult> {
  return post<DCWriteResult>('write-file', { path, content, mode });
}

export async function editBlock(
  filePath: string,
  oldString: string,
  newString: string,
): Promise<{ success: boolean; description?: string }> {
  return post('edit-block', { file_path: filePath, old_string: oldString, new_string: newString });
}

export async function listDirectory(path: string, depth = 1): Promise<DCListResult> {
  return post<DCListResult>('list-directory', { path, depth });
}

export async function createDirectory(path: string): Promise<{ success: boolean; description?: string }> {
  return post('create-directory', { path });
}

export async function moveFile(
  source: string,
  destination: string,
): Promise<{ success: boolean; description?: string }> {
  return post('move-file', { source, destination });
}

export async function getFileInfo(path: string): Promise<DCFileInfo> {
  return post<DCFileInfo>('get-file-info', { path });
}

export async function searchFiles(
  pathOrQuery: string,
  queryOrPath?: string,
  maxResults = 20,
): Promise<DCSearchResult> {
  const looksLikePath = /^[A-Za-z]:[\\/]/.test(pathOrQuery) || pathOrQuery.includes('/') || pathOrQuery.includes('\\');
  const path = looksLikePath ? pathOrQuery : queryOrPath;
  const query = looksLikePath ? (queryOrPath || '') : pathOrQuery;
  return post<DCSearchResult>('search-files', { query, path, max_results: maxResults });
}

export async function executeCommand(
  command: string,
  options: { shell?: string; timeout_ms?: number; cwd?: string } = {},
): Promise<DCCommandResult> {
  return post<DCCommandResult>('execute-command', {
    command,
    shell: options.shell,
    timeout_ms: options.timeout_ms,
    cwd: options.cwd,
  });
}

export async function getSystemInfo(): Promise<DCSystemInfoResult> {
  return get<DCSystemInfoResult>('system-info');
}

export async function readMultipleFiles(
  paths: string[],
): Promise<{ path: string; result?: DCFileResult; error?: string }[]> {
  const results = await Promise.allSettled(paths.map((path) => readFile(path)));
  return paths.map((path, index) => {
    const result = results[index];
    if (result.status === 'fulfilled') {
      return { path, result: result.value };
    }
    return { path, error: (result.reason as Error).message };
  });
}
